File: api-dot-net/Controllers/CodeExecutionController.cs

using System.Security.Claims;
using System.Threading.Tasks;
using CjsApi.Dto.RequestDto;
using CjsApi.Dto.ResponseDto;
using CjsApi.Services;
using CjsApi.Services.CodeExecution.Dto;
using Microsoft.AspNetCore.Authorization;
using Microsoft.AspNetCore.Mvc;

namespace CjsApi.Controllers
{

    [ApiController]
    [Route("api/code")]
    public class CodeExecutionController : ControllerBase
    {

        private readonly CodeExecutionService _service;
        private readonly SubmissionService _submissionService;

        public CodeExecutionController(CodeExecutionService service, SubmissionService submissionService)
        {
            _service = service;
            _submissionService = submissionService;
        }

        [Authorize]
        [HttpPost("run")]
        public async Task<ActionResult<ApiResponseDto<string>>> RunCode(
        [FromBody] CodeRunRequestDto dto)
        {
            try
            {
                var jobId = await _service.RunAsync(dto);


                return Ok(new ApiResponseDto<string>(
                    true,
                    "Execution started",
                    jobId
                ));
            }
            catch (Exception e)
            {
                return BadRequest(new ApiResponseDto<string>(
                                    false,
                                    e.Message,
                                    null
                                ));

            }
        }


        [Authorize]
        [HttpPost("submit")]
        public async Task<ActionResult<ApiResponseDto<string>>> SubmitCode(
       [FromBody] CodeRunRequestDto dto)
        {
            try
            {


                var userIdClaim = User.FindFirstValue(ClaimTypes.NameIdentifier);

                if (!int.TryParse(userIdClaim, out int userId))
                    return Unauthorized("Invalid user id in token");


                var s = await _submissionService.CreateSubmissionAsync(userId, dto.ProblemId, dto.SourceCode, dto.Language);
                var jobId = await _service.SubmitAsync(dto, s.Id);

                return Ok(new ApiResponseDto<string>(
                    true,
                    "Execution started",
                    jobId
                ));

            }
            catch (Exception e)
            {


                return Ok(new ApiResponseDto<string>(
                    true,
                    "Execution started",
                    e.Message
                ));
            }
        }

        // 2️⃣ Poll status
        [HttpGet("status/{jobId}")]
        public ActionResult<ApiResponseDto<object>> GetStatus(string jobId)
        {
            var job = _service.GetStatus(jobId);

            if (job == null)
                return NotFound();
            return Ok(new ApiResponseDto<ExecutionJob>(
                true,
                "Status fetched",
                job
            ));
        }

    }
}

File: java-api/cjs-service/src/main/java/com/cjs/cjs_service/controller/CodeExecutionController.java
package com.cjs.cjs_service.controller;

import com.cjs.cjs_service.dto.request.CodeRunRequestDto;
import com.cjs.cjs_service.dto.response.ApiResponseDto;
import com.cjs.cjs_service.model.Submission;
import com.cjs.cjs_service.service.SubmissionService;
import com.cjs.cjs_service.service.codeExecutionSerivce.CodeExecutionService;
import com.cjs.cjs_service.dto.ExecutionJob;

import org.springframework.http.ResponseEntity;

import org.springframework.security.core.Authentication;

import org.springframework.web.bind.annotation.*;

@RestController
@RequestMapping("/api/code")
public class CodeExecutionController {

    private final CodeExecutionService service;
    private final SubmissionService submissionService;

    public CodeExecutionController(
            CodeExecutionService service,
            SubmissionService submissionService) {
        this.service = service;
        this.submissionService = submissionService;
    }

    // ===============================
    // Run Code
    // ===============================
    // @PreAuthorize("isAuthenticated()")
    @PostMapping("/run")
    public ResponseEntity<ApiResponseDto<String>> runCode(
            @RequestBody CodeRunRequestDto dto) {

        try {
            String jobId = service.run(dto);

            return ResponseEntity.ok(
                    new ApiResponseDto<String>(
                            "Execution started",
                            jobId,
                            true

                    ));
        } catch (Exception e) {
            return ResponseEntity
                    .badRequest()
                    .body(new ApiResponseDto<>(
                            e.getMessage(),
                            null,
                            false));
        }
    }

    // ===============================
    // Submit Code
    // ===============================
    // @PreAuthorize("isAuthenticated()")
    @PostMapping("/submit")
    public ResponseEntity<ApiResponseDto<String>> submitCode(
            @RequestBody CodeRunRequestDto dto,
            Authentication authentication) {

        try {
            // Extract user id from JWT
            // Jwt jwt = (Jwt) authentication.getPrincipal();
            // Integer userId = jwt.getClaim("sub") != null
            // ? Integer.parseInt(jwt.getClaim("sub"))
            // : null;

            // if (userId == null) {
            // return ResponseEntity
            // .status(HttpStatus.UNAUTHORIZED)
            // .body(new ApiResponseDto<>(
            // "Invalid user id in token",
            // null,
            // false));
            // }

            Submission submission = submissionService.createSubmission(
                    1,
                    dto.getProblemId(),
                    dto.getCode(),
                    dto.getLanguage());

            String jobId = service.submit(dto, submission.getId());

            return ResponseEntity.ok(
                    new ApiResponseDto<>(
                            "Execution started",
                            jobId,
                            true

                    ));

        } catch (Exception e) {
            // ⚠️ Kept behavior same as your .NET code
            return ResponseEntity.ok(
                    new ApiResponseDto<>(
                            "Submission failed: " + e.getMessage(),
                            null,
                            false

                    ));
        }
    }

    // ===============================
    // Get Status
    // ===============================
    @GetMapping("/status/{jobId}")
    public ResponseEntity<ApiResponseDto<ExecutionJob>> getStatus(
            @PathVariable String jobId) {

        ExecutionJob job = service.getStatus(jobId);

        if (job == null) {
            return ResponseEntity.notFound().build();
        }

        return ResponseEntity.ok(
                new ApiResponseDto<>(
                        "Status fetched",
                        job,
                        true

                ));
    }
}
